tyrian/js/src/main/scala/tyrian/websocket/WebSocket.scala

```diff
@@ -4,85 +4,110 @@ import org.scalajs.dom
       address: String,
-      onOpenSendMessage: Option[String]
-  ): Either[WebSocketEvent.Error, LiveSocket] =
-    try {
-      val socket = new dom.WebSocket(address)
+      onOpenSendMessage: Option[String],
+      settings: KeepAliveSettings
+  ): Task[String, LiveSocket] =
+    Task.Delay { () =>
+      val socket    = new dom.WebSocket(address)
+      val keepAlive = new KeepAlive(socket, settings.message, settings.timeout)
 
       val subs =
         Sub.Batch(
           Sub.fromEvent("message", socket) { e =>
             Some(WebSocketEvent.Receive(e.asInstanceOf[dom.MessageEvent].data.toString))
           },
-          Sub.fromEvent("error", socket)(_ => Some(WebSocketEvent.Error("Web socket connection error"))),
-          Sub.fromEvent("close", socket)(e => Some(WebSocketEvent.Close)),
-          Sub.fromEvent("open", socket) { e =>
+          Sub.fromEvent("error", socket) { e =>
+            val msg =
+              try { e.asInstanceOf[dom.ErrorEvent].message }
+              catch { case _: Throwable => "Unknown" }
+            Some(WebSocketEvent.Error(msg))
+          },
+          Sub.fromEvent("close", socket) { e =>
+            if settings.enabled then keepAlive.cancel() else ()
+            val ev = e.asInstanceOf[dom.CloseEvent]
+            Some(WebSocketEvent.Close(ev.code, ev.reason))
+          },
+          Sub.fromEvent("open", socket) { _ =>
             onOpenSendMessage.foreach(msg => socket.send(msg))
+            if settings.enabled then keepAlive.run() else ()
             Some(WebSocketEvent.Open)
           }
         )
 
-      Right(LiveSocket(socket, subs))
-    } catch {
-      case e: Throwable =>
-        Left(
-          WebSocketEvent.Error(s"Error trying to set up websocket: ${e.getMessage}")
-        )
+      LiveSocket(socket, subs)
     }
 
-  final class LiveSocket(val socket: dom.WebSocket, val subs: Sub[WebSocketEvent])
-
-  enum WebSocketReadyState derives CanEqual:
-    case CONNECTING, OPEN, CLOSING, CLOSED
-
-    def isOpen: Boolean =
-      this match
-        case CLOSED  => false
-        case CLOSING => false
-        case _       => true
-
-  object WebSocketReadyState:
-    def fromInt(i: Int): WebSocketReadyState =
-      i match {
-        case 0 => CONNECTING
-        case 1 => OPEN
-        case 2 => CLOSING
-        case 3 => CLOSED
-        case _ => CLOSED
-      }
-
-object WebSocket:
+  final class KeepAlive(socket: dom.WebSocket, msg: String, timeout: FiniteDuration):
+    @SuppressWarnings(Array("scalafix:DisableSyntax.var"))
+    private var timerId = 0;
 
-  def apply(address: String): WebSocket =
-    new WebSocket(address, None)
+    @SuppressWarnings(Array("scalafix:DisableSyntax.null"))
+    def run(): Unit =
+      if socket != null && WebSocketReadyState.fromInt(socket.readyState).isOpen then
+        println("[info] Sending heartbeat 💓")
+        socket.send(msg)
+      timerId = dom.window.setTimeout(Functions.fun0(() => run()), timeout.toMillis.toDouble)
```

Comment:
No need to change it now, but I wonder if we can do this with a `Sub`? I'll have a look after it's merged.